mesonbuild/modules/unstable_wayland.py

```diff
@@ -0,0 +1,121 @@
+        self.pkgdatadir = None
+        self.scanner_bin = None
+
+        self.methods.update({
+            'scan_xml': self.scan_xml,
+            'find_protocol': self.find_protocol,
+        })
+
+    @typed_pos_args('wayland.scan_xml', varargs=(str, File), min_varargs=1)
+    @typed_kwargs(
+        'wayland.scan_xml',
+        KwargInfo('side', str, default='client', validator=in_set_validator({'client', 'server'})),
+        KwargInfo('scope', str, default='private', validator=in_set_validator({'private', 'public'})),
+    )
+    def scan_xml(self, state, args, kwargs):
+        if self.scanner_bin is None:
+            self.scanner_bin = state.find_program('wayland-scanner', for_machine=MachineChoice.BUILD)
+
+        scope = kwargs['scope']
+        side = kwargs['side']
+
+        xml_files = self.interpreter.source_strings_to_files(args[0])
+        targets = []
+        for xml_file in xml_files:
+            base_name = os.path.basename(xml_file.fname)
+            name, _ = os.path.splitext(base_name)
+
+            code = CustomTarget(
+                f'{name}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{scope}-code', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-protocol.c'],
+                backend=state.backend,
+            )
+            targets.append(code)
+
+            header = CustomTarget(
+                f'{name}-{side}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{side}-header', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-{side}-protocol.h'],
+                backend=state.backend,
+            )
+            targets.append(header)
+
+        return ModuleReturnValue(targets, targets)
+
+    @typed_pos_args('wayland.find_protocol', str)
+    @typed_kwargs(
+        'wayland.find_protocol',
+        KwargInfo('state', str, default='stable', validator=in_set_validator({'stable', 'staging', 'unstable'})),
+        KwargInfo('version', (int, NoneType)),
+    )
+    def find_protocol(self, state, args, kwargs):
+        base_name = args[0]
+        xml_state = kwargs['state']
+        version = kwargs['version']
+
+        if (xml_state == 'staging' or xml_state == 'unstable') and version is None:
+            raise MesonException(f'{xml_state} protocols require a version number.')
+
+        if self.protocols_dep is None:
+            self.protocols_dep = self.interpreter.func_dependency(state.current_node, ['wayland-protocols'], {})
+
+        if self.pkgdatadir is None:
+            self.pkgdatadir = self.protocols_dep.get_variable(pkgconfig='pkgdatadir', internal='pkgdatadir')
+
+        if xml_state == 'stable':
+            if version is not None:
+                mlog.warning('stable protocols do not require a version number.', location=self.interpreter.current_node, once=True)
```

Comment:
I would prefer a hard error.

Reply:
Also this validation check could be moved up above to where we check and error out on a missing one for unstable/staging, as it's better to error out or warn on invalid arguments earlier rather than later, and currently it happens after checking the environment state (dependency lookup).